temporal-sdk/src/main/java/io/temporal/worker/tuning/ResourceBasedController.java

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright (C) 2022 Temporal Technologies, Inc. All Rights Reserved.
+ *
+ * Copyright (C) 2012-2016 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Modifications copyright (C) 2017 Uber Technologies, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this material except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.temporal.worker.tuning;
+
+import com.uber.m3.tally.Gauge;
+import com.uber.m3.tally.Scope;
+import io.temporal.common.Experimental;
+import io.temporal.worker.MetricsType;
+import java.time.Instant;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.concurrent.locks.ReentrantLock;
+
+/**
+ * Is used by {@link ResourceBasedSlotSupplier} and {@link ResourceBasedTuner} to make decisions
+ * about whether slots should be handed out based on system resource usage.
+ */
+@Experimental
+public class ResourceBasedController {
+  public final ResourceBasedControllerOptions options;
+
+  private final ReentrantLock decisionLock = new ReentrantLock();
+  private final PIDController memoryController;
+  private final PIDController cpuController;
+  private final SystemResourceInfo systemInfoSupplier;
+  private Instant lastPidRefresh = Instant.now();
+
+  private AtomicReference<Metrics> metrics = new AtomicReference<>();
+  private boolean metricsScopeSet;
+  private volatile double lastMemUsage;
+  private volatile double lastCpuUsage;
```

Comment:
Why are you using volatile and a lock. The lock should be enough to prevent a race.

Also it seems like these values are written to, but never read?

Reply:
Ah, yeah they're not any more. Missed that when I changed the metrics stuff.

They needed to be volatile before because they were *read* outside the lock, but now I'll just get rid of them.